readthedocs/gold/views.py

```diff
@@ -152,3 +139,149 @@ def post(self, request, *args, **kwargs):
+
+        scheme = 'https' if settings.PUBLIC_DOMAIN_USES_HTTPS else 'http'
+        return_url = f'{scheme}://{settings.PRODUCTION_DOMAIN}' + str(self.get_success_url())
+        try:
+            billing_portal = stripe.billing_portal.Session.create(
+                customer=stripe_customer,
+                return_url=return_url,
+            )
+            return HttpResponseRedirect(billing_portal.url)
+        except:  # noqa
+            messages.error(
+                request,
+                _('There was an error connecting to Stripe, please try again in a few minutes'),
+            )
+            return HttpResponseRedirect(self.get_success_url())
+
+    def get_success_url(self):
+        return reverse('gold_detail')
+
+
+# TODO: where this code should live? (readthedocs-ext?)
+# Should we move all GoldUser code to readthedocs-ext?
+class StripeEventView(APIView):
+
+    """Endpoint for Stripe events."""
+
+    permission_classes = (permissions.AllowAny,)
+    renderer_classes = (JSONRenderer,)
+
+    EVENT_CHECKOUT_PAYMENT_FAILED = 'checkout.session.async_payment_failed'
+    EVENT_CHECKOUT_PAYMENT_SUCCEEDED = 'checkout.session.async_payment_succeeded'
+    EVENT_CHECKOUT_COMPLETED = 'checkout.session.completed'
+    EVENT_CUSTOMER_SUBSCRIPTION_UPDATED = 'customer.subscription.updated'
+
+    EVENTS = [
+        EVENT_CHECKOUT_PAYMENT_SUCCEEDED,
+        EVENT_CHECKOUT_COMPLETED,
+        EVENT_CUSTOMER_SUBSCRIPTION_UPDATED,
+    ]
+
+    def post(self, request, format=None):
+        try:
+            event = stripe.Event.construct_from(request.data, settings.STRIPE_SECRET)
+            if event.type in self.EVENTS:
+                stripe_customer = event.data.object.customer
+
+                if event.type == self.EVENT_CHECKOUT_COMPLETED:
+                    username = event.data.object.client_reference_id
+                    subscription = stripe.Subscription.retrieve(event.data.object.subscription)
+
+                    user = User.objects.get(username=username)
+                    GoldUser.objects.create(
+                        user=user,
+                        level=subscription.plan.id,
+                        stripe_id=stripe_customer,
+                        subscribed=True,
+                    )
+                    # TODO: add user notification saying it was successful
+
+                elif event.type == self.EVENT_CHECKOUT_PAYMENT_FAILED:
+                    username = event.data.object.client_reference_id
+                    # TODO: add user notification saying it failed
+                    log.exception(
+                        'Gold User payment failed. username=%s customer=%s',
+                        username,
+                        stripe_customer,
+                    )
+
+                elif event.type == self.EVENT_CUSTOMER_SUBSCRIPTION_UPDATED:
+                    # TODO: check if the subscription was canceled, past due,
+                    # etc and take the according action
```

Comment:
Yea, that seems useful. I believe we haven basically just never cancelled a Gold sub, even when payments stopped. Would be useful to do more state tracking if it isn't too difficult.

Reply:
Shouldn't be too difficult. Besides, we already have this code written for .com --I want to bring `subscriptions` app into .org eventually and share all this code if possible.

Reply:
Yea, that would be ideal, it seems.